packages/relay/src/lib/eth.ts

```diff
@@ -144,14 +150,79 @@ export class EthImpl implements Eth {
       return feeHistory;
     } catch (e) {
       this.logger.error(e, 'Error constructing default feeHistory');
+      return {
+        baseFeePerGas: [],
+        gasUsedRatio: [],
+        reward: [],
+        oldestBlock: EthImpl.zeroHex
+      };
     }
   }
 
-  private async getFeeWeibars() {
+  private async getFeeByBlockNumber(blockNumber: number): Promise<string> {
+    let fee = 0;
+
+    try {
+      const block = await this.mirrorNodeClient.getBlock(blockNumber);
+      fee = await this.getFeeWeibars(`lte:${block.timestamp.to}`);
+    } catch (error) {
+      this.logger.warn(error, `Fee history cannot retrieve block or fee. Returning ${fee} fee for block ${blockNumber}`);
+    }
+
+    return EthImpl.numberTo0x(fee);
+  }
+
+  private async getFeeHistory(blockCount: number, newestBlockNumber: number, latestBlockNumber: number, rewardPercentiles: Array<number> | null) {
+    blockCount = blockCount > constants.MAX_FEE_HISTORY_RESULTS ? constants.MAX_FEE_HISTORY_RESULTS : blockCount;
+
+    if (blockCount === 0) {
+      return {
+        gasUsedRatio: null,
+        oldestBlock: EthImpl.zeroHex
+      };
+    }
+
+    // include newest block number in the total block count
+    const oldestBlockNumber = Math.max(0, newestBlockNumber - blockCount + 1);
+    const shouldIncludeRewards = Array.isArray(rewardPercentiles) && rewardPercentiles.length > 0;
+    const feeHistory = {
+      baseFeePerGas: [] as string[],
+      gasUsedRatio: [] as string[],
+      oldestBlock: EthImpl.numberTo0x(oldestBlockNumber),
+    };
+
+    // get fees from oldest to newest blocks
+    for(let blockNumber = oldestBlockNumber; blockNumber <= newestBlockNumber; blockNumber++) {
+      const fee = await this.getFeeByBlockNumber(blockNumber);
+
+      feeHistory.baseFeePerGas.push(fee);
+      feeHistory.gasUsedRatio.push(EthImpl.defaultGasUsedRatio);
+    }
+
+    // get latest block fee
+    let nextBaseFeePerGas = _.last(feeHistory.baseFeePerGas);
+
+    if (latestBlockNumber > newestBlockNumber) {
+      // get next block fee if the newest block is not the latest
+      nextBaseFeePerGas = await this.getFeeByBlockNumber(newestBlockNumber + 1);
+    }
+    
+    if (nextBaseFeePerGas) {
+      feeHistory.baseFeePerGas.push(nextBaseFeePerGas);
```

Comment:
will this result in the length of `baseFeePerGas` and `gasUsedRatio` being different?
Should `gasUsedRatio` be updated also?

Reply:
Yes, the length of both arrays is different by 1 element. The `baseFeePerGas` array always includes the next block after the newest but `gasUsedRatio` does not.

Reply:
Okay